From 28af3ea1eccbfc476cc60d9163e787c2963342b2 Mon Sep 17 00:00:00 2001
From: "Reece R. Pollack" <x0183204@ti.com>
Date: Wed, 4 Feb 2015 17:40:56 -0500
Subject: [PATCH 240/256] net: Fix skb->protocol setting for mixed IPsec
 tunnels

When processing fragmented packets in a mixed-protocol tunnel
(i.e. IPv6 inner and IPv4 outer, or IPv4 inner and IPv6 outer),
the skb->protocol field is being set to the wrong value. This
patch sets the skb->protocol value according to the IP header
version of the packet it's carrying.

Signed-off-by: Reece R. Pollack <x0183204@ti.com>
(cherry picked from commit 251e0840063b3eafb3905e5d4f6db4b6da84ceb5)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 net/ipv4/xfrm4_output.c |    9 ++++++++-
 net/ipv6/xfrm6_output.c |    9 ++++++++-
 2 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/net/ipv4/xfrm4_output.c b/net/ipv4/xfrm4_output.c
index baa0f63..62c5a47 100644
--- a/net/ipv4/xfrm4_output.c
+++ b/net/ipv4/xfrm4_output.c
@@ -73,6 +73,8 @@ EXPORT_SYMBOL(xfrm4_prepare_output);
 
 int xfrm4_output_finish(struct sk_buff *skb)
 {
+	const struct iphdr *iph;
+
 #ifdef CONFIG_NETFILTER
 	if (!skb_dst(skb)->xfrm) {
 		IPCB(skb)->flags |= IPSKB_REROUTED;
@@ -82,7 +84,12 @@ int xfrm4_output_finish(struct sk_buff *skb)
 	IPCB(skb)->flags |= IPSKB_XFRM_TRANSFORMED;
 #endif
 
-	skb->protocol = htons(ETH_P_IP);
+	iph = ip_hdr(skb);
+	if (iph->version == 6)
+		skb->protocol = htons(ETH_P_IPV6);
+	else
+		skb->protocol = htons(ETH_P_IP);
+
 	return xfrm_output(skb);
 }
 
diff --git a/net/ipv6/xfrm6_output.c b/net/ipv6/xfrm6_output.c
index 6cd625e..f591507 100644
--- a/net/ipv6/xfrm6_output.c
+++ b/net/ipv6/xfrm6_output.c
@@ -128,11 +128,18 @@ EXPORT_SYMBOL(xfrm6_prepare_output);
 
 int xfrm6_output_finish(struct sk_buff *skb)
 {
+	const struct ipv6hdr *iphv6;
+
 #ifdef CONFIG_NETFILTER
 	IP6CB(skb)->flags |= IP6SKB_XFRM_TRANSFORMED;
 #endif
 
-	skb->protocol = htons(ETH_P_IPV6);
+	iphv6 = ipv6_hdr(skb);
+	if (iphv6->version == IPVERSION)
+		skb->protocol = htons(ETH_P_IP);
+	else
+		skb->protocol = htons(ETH_P_IPV6);
+
 	return xfrm_output(skb);
 }
 
-- 
1.7.5.4

